proto/p4/p4runtime.proto

```diff
@@ -334,8 +342,20 @@ message StreamMessageResponse {
 // Packet sent from the switch to the controller.
 message PacketIn {
   bytes payload = 1;
-  uint32 ingress_physical_port = 2;
-  uint32 ingress_logical_port = 3;
+  // This will be based on P4 header annotated as
+  // @controller_header("packet_in").
+  // At most one P4 header can have this annotation.
+  repeated PacketMetadata metadata = 2;
+
+  // TODO: Remove fields below once users of PacketIn have migrated to
+  // PacketMetadata.
+  uint32 ingress_physical_port = 3;
+  uint32 ingress_logical_port = 4;
+}
+
+message PacketMetadata {
+  uint32 metadata_id = 1;
```

Comment:
Please add a comment to explain the metadata_id.  Does it refer to the ID in the P4Info metadata fields in ControllerPacketMetadata?